layers/mdp.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2024 Google, Inc. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file in the root of the source
+// tree.
+
+package layers
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+
+	"github.com/gopacket/gopacket"
+)
+
+const (
+	MdpTlvType uint8 = iota
+	MdpTlvLength
+	MdpTlvDeviceInfo
+	MdpTlvNetworkInfo
+	MdpTlvLongitude
+	MdpTlvLatitude
+	MdpTlvType6
+	MdpTlvType7
+	MdpTlvIP          = 11
+	MdpTlvUnknownBool = 13
+	MdpTlvEnd         = 255
+)
+
+// MDP defines a MDP over LLC layer.
+type MDP struct {
+	BaseLayer
+	PreambleData []byte
+	DeviceInfo   string
+	NetworkInfo  string
+	Longitude    float64
+	Latitude     float64
+	Type6UUID    string
+	Type7UUID    string
+	IPAddress    net.IP
+	Type13Bool   bool
+
+	Type   EthernetType
+	Length int
+}
+
+// LayerType returns LayerTypeMDP.
+func (m *MDP) LayerType() gopacket.LayerType { return LayerTypeMDP }
+
+// DecodeFromBytes decodes the given bytes into this layer.
+func (m *MDP) DecodeFromBytes(data []byte, df gopacket.DecodeFeedback) error {
+	var length int
+	if len(data) < 28 {
+		df.SetTruncated()
+		return fmt.Errorf("MDP length %d too short", len(data))
+	}
+	m.Type = EthernetTypeMerakiDiscoveryProtocol
+	m.Length = len(data)
+	offset := 28
+	m.PreambleData = data[:offset]
+
+	for {
+		if offset >= m.Length {
+			break
+		}
+		t := data[offset]
+		switch t {
+		case MdpTlvDeviceInfo:
+			offset += 2
+			length = int(data[offset-1])
+			m.Contents = append(m.Contents, data[offset-2:offset+length]...)
+			m.DeviceInfo = string(data[offset : offset+length])
+			offset += length
+			break
+		case MdpTlvNetworkInfo:
+			offset += 2
+			length = int(data[offset-1])
+			m.NetworkInfo = string(data[offset : offset+length])
+			offset += length
+			break
+		case MdpTlvLongitude:
+			offset += 2
+			length = int(data[offset-1])
+			m.Longitude, _ = strconv.ParseFloat(string(data[offset:offset+length]), 64)
+			offset += length
+			break
+		case MdpTlvLatitude:
+			offset += 2
+			length = int(data[offset-1])
+			m.Latitude, _ = strconv.ParseFloat(string(data[offset:offset+length]), 64)
+			offset += length
+			break
+		case MdpTlvType6:
+			offset += 2
+			length = int(data[offset-1])
+			m.Type6UUID = string(data[offset : offset+length])
+			offset += length
+			break
+		case MdpTlvType7:
+			offset += 2
+			length = int(data[offset-1])
+			m.Type7UUID = string(data[offset : offset+length])
+			offset += length
+			break
+		case MdpTlvIP:
+			offset += 2
+			length = int(data[offset-1])
+			m.IPAddress = net.ParseIP(string(data[offset : offset+length]))
+			offset += length
+			break
+		case MdpTlvUnknownBool:
+			offset += 2
+			length = int(data[offset-1])
+			m.Type13Bool, _ = strconv.ParseBool(string(data[offset : offset+length]))
```

Comment:
where does this come from? I can't see something similar in Wireshark's code.

Reply:
`MdpTlvIP` and `MdpTlvUnknownBool` are two types I saw when capturing MDP from a WiFi AP. The types and contents were obvious based on the ASCII.

AFAIK this protocol isn't officially documented so these types (and others) may have been added to the protocol since the Wireshark dissector was created.